src/Microsoft.AspNet.Security/Infrastructure/SecurityHelper.cs

```diff
@@ -22,20 +22,38 @@ public static class SecurityHelper
         /// <param name="identity"></param>
         public static void AddUserIdentity([NotNull] HttpContext context, [NotNull] IIdentity identity)
         {
-            var newClaimsPrincipal = new ClaimsPrincipal(identity);
-
-            ClaimsPrincipal existingPrincipal = context.User;
-            if (existingPrincipal != null)
+            if (context.User == null)
+            {
+                context.User = new ClaimsPrincipal(identity);
+            }
+            else
             {
-                foreach (var existingClaimsIdentity in existingPrincipal.Identities)
+                ClaimsIdentity claimsIdentity = identity as ClaimsIdentity;
+                if (claimsIdentity == null)
+                {
+                    context.User.AddIdentity(new ClaimsIdentity(identity));
+                }
+                else
                 {
-                    if (existingClaimsIdentity.IsAuthenticated)
```

Comment:
The IsAuthenticated check here was being used to filter out the default anonymous identity.